### PR TITLE
ramips: add support for ZyXEL Keenetic Giga III

### PR DESCRIPTION
**Pull request by remittor**
**ZyXEL Keenetic Giga III** is a 2.4/5 GHz band 11ac router, based on MT7621ST.

Specification:
-------------
* SoC: MediaTek MT7621ST (1C/2T)
* RAM: 256 MB DDR3
* Flash: 128 MB NAND S34ML01G200TFI00
* Ethernet: 5x 10/100/1000 Mbps
* WIFI: 5 GHz   MT7612EN nac
* WIFI: 2.4 GHz MT7602EN bgn
* USB: 1x 2.0, 1x 3.0
* BTN: Power, Reset, WPS, FN1, FN2
* LEDS: Power(Green),Fn(Green),Wan(Green),WiFi2(Green),WiFi5(Green),USB1(Green),USB2(Green)
* UART: present as five pads without through-holes on the PCB.
Pads are located above 2.4G LED and under board logotype (GND/empty/RX/TX/Vcc).
UART uses 3.3V and settings: 57600-8-N-1

MAC addresses as verified by stock firmware:
----------------------------------------------

| Interface   |       MAC         | Factory | Format |
|-------------|-------------------|---------|--------|
| WiFi (2.4g) | xx:xx:xx:xx:xx:20 | 0x4     | binary |
| LAN         | xx:xx:xx:xx:xx:20 | 0x4     | binary |
| WAN (label) | xx:xx:xx:xx:xx:21 | 0x28    | binary |
| WiFi (5g)   | xx:xx:xx:xx:xx:22 | 0x8004  | binary |

Installation via U-Boot:
------------------------
* Download the latest OpenWrt firmware factory-image and rename it to kgiga3_recovery.bin
* Set up a Tftp server on a PC (e.g. Tftpd32) and place the firmware image to the root directory of the server.
* Power off the router and use a twisted pair cable to connect the PC to any of the router's LAN ports.
* Configure the network adapter of the PC to use IP address 192.168.1.2 and subnet mask 255.255.255.0.
* Power up the router while holding the reset button pressed.
* Wait approximately for 5 seconds and then release the reset button.
* The router should download the firmware via TFTP and complete flashing in a few minutes.
* Connect with SSH to 192.168.1.1 and set a root password or browse to http://192.168.1.1 if LuCI is installed.

Installation via Keenetic-OS GUI:
------------------------------------------------
* Set the IP address 192.168.1.1 for the connected device (in the KeeneticOS settings and in the network connection properties).
* Open the page "http://192.168.1.1/a" in the browser, enter the command "more proc:/dual_image/boot_active" (without quotes) in the "Command" field and click the "Send request" button.
* A response will appear in the browser, from which you need to select only the value of the "message" parameter.
* If the value of the "message" parameter is not equal to 2, then you need to force flash any official firmware and go to the first step of the instruction.
* To install OpenWRT, go to the "General Settings" page and in the "System Files" section, click on "firmware".
* Then click the "Replace File" button.
* In the window that appears, specify the location of the file "openwrt-ramips-mt7621-zyxel_keenetic-giga-iii-squashfs-factory.bin".
* The installation of OpenWRT on the device will begin.
* After installing the firmware, the device will reboot itself and start OpenWRT (takes about 60...90 seconds).

Revert to the stock firmware via OpenWrt GUI:
------------------------------------------------
* Upload the stock image using "Flash Firmware" interface.
* Select "Force update" and continuing upgrade.

### CODE REVIEW

**Comment by jospezial** [comment]
Telnet? SSH would be better.

**Comment by remittor** [comment]
> Telnet? SSH would be better.

See: https://openwrt.org/toh/zyxel/zyxel_keenetic_omni_rev._a#oem_easy_installation
And many other manuals.

**Comment by jospezial** [comment]
The page seems to be not up to date.
It looks as telnet(d) is not in the openwrt images.
https://github.com/openwrt/openwrt/commit/d196b1fc2eddf7cbab4351452b6fc9b41f15b142

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 1]
```diff
@@ -0,0 +1,301 @@
+#include "mt7621.dtsi"
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:1]: Please add an SPDX license identifier.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+		label-mac-device = &gmac0;
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,57600n8";
+	};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:20]: This block can be dropped, it's in the DTSI.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+	keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:26]: Use gpio-keys instead. Just look at any other DTS file for mt7621.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			/* debounce-interval = <60>; */
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:32]: Remove that comment.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+			label = "fn1";
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+		};
+
+		fn2 {
+			label = "fn2";
+			gpios = <&gpio 10 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_1>;
+		};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:51]: These blocks should be sorted by something. Either GPIO number or name.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+			gpios = <&gpio 9 GPIO_ACTIVE_HIGH>;
+			default-state = "keep";
+		};
+
+		fn {
+			label = "green:fn";
+			gpios = <&gpio 12 GPIO_ACTIVE_HIGH>;
+		};
+
+		internet { /* WAN */
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:68]: Remove the comment.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+		devices = <&ubiconcat0 &ubiconcat1>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "ubi";         // special name for creating UBIFS
+				reg = <0x0 0x7500000>; // 117 MB
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:128]: Remove comments

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+			};
+		};
+	};
+};
+
+&nand {
+	status = "okay";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:135]: Empty line after status.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x80000>; // 512 KB
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:143]: Again, remove comments. Also all below, I won't comment again ...

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+		factory: partition@100000 {
+			label = "factory";        // RF-EEPROM
+			reg = <0x100000 0x80000>; // 512 KB
+			read-only;
+		};
+
+		/* U-Boot expects to find kernel image at 0x180000 */
+		partition@180000 {
+			label = "firmware";         // Firmware_1
+			reg = <0x180000 0x3E40000>; // 122 MB
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:162]: Please use lower case for hex letters.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+		partition@40c0000 {
+			label = "factory-res";     // RF-EEPROM_res
+			reg = <0x40c0000 0x80000>; // 512 KB
+			read-only;
+		};
+
+		ubiconcat1: partition@4140000 {
+			label = "ubiconcat1";        // Firmware_2
+			reg = <0x4140000 0x3AC0000>; // 58 MB
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:200]: lower case.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+		};
+
+		/* U-Boot used latest 32 BLOCK for factory bbt table */
+		partition@7C00000 { // 128MiB - 32*128KiB = 0x7C00000
+			label = "mtk-nand-bbt";
+			reg = <0x7C00000 0x400000>; // 4 MB
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:206]: lower case.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+			read-only;
+		};
+
+		ubiconcat1: partition@4140000 {
+			label = "ubiconcat1";        // Firmware_2
+			reg = <0x4140000 0x3AC0000>; // 58 MB
+		};
+
+		/* U-Boot used latest 32 BLOCK for factory bbt table */
+		partition@7C00000 { // 128MiB - 32*128KiB = 0x7C00000
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:204]: lower case.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+			label = "mtk-nand-bbt";
+			reg = <0x7C00000 0x400000>; // 4 MB
+			read-only;
+		};
+	};
+};
+
+&xhci {
+	status = "okay";
+};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:214]: This can be dropped, xhci is enabled by default.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+
+&pcie0 { // MT7612EN wifi5g
+	wifi0: mt76@0,0 {  
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x8000>;
+		nvmem-cells = <&macaddr_factory_8004>;
+		nvmem-cell-names = "mac-address";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:226]: nvmem-cells can be dropped here, this location is loaded automatically from mtd-eeprom.

https://openwrt.org/docs/guide-developer/mac.address#mac_address_pulled_by_driver

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+&xhci {
+	status = "okay";
+};
+
+&pcie {
+	status = "okay";
+};
+
+&pcie0 { // MT7612EN wifi5g
+	wifi0: mt76@0,0 {
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:221]: wifi0: and wifi1: DT labels are unused and can be dropped.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+	};
+};
+
+&pcie1 { // MT7602EN wifi2g
+	wifi1: mt76@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+		nvmem-cells = <&macaddr_factory_4>;
+		nvmem-cell-names = "mac-address";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:237]: Like above, nvmem-cell* can be dropped.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+		ieee80211-freq-limit = <2400000 2500000>;
+	};
+};
+
+&gmac0 {
+	nvmem-cells = <&macaddr_factory_28>;
+	nvmem-cell-names = "mac-address";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:244]: This does not match your description in commit message.

While it's uncommon anyway (typically LAN is 0x28 and WAN is 0x2e), based on your description I'd expect 0x4 here.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+		reg = <0x4 0x6>;
+	};
+
+	macaddr_factory_28: macaddr@28 { /* WAN mac (LABEL) */
+		reg = <0x28 0x6>;
+	};
+
+	macaddr_factory_8004: macaddr@8004 { /* wifi5 mac (wan + 1) */
+		reg = <0x8004 0x6>;
+	};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:300]: This last block can be dropped, since we get it from mtd-eeprom.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+			label = "wan";
+			nvmem-cells = <&macaddr_factory_28>;
+			nvmem-cell-names = "mac-address";
+		};
+	};
+};
+
+&state_default {
+	gpio {
+		groups = "uart3", "jtag", "wdt";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:280]: On a quick look, uart2 seems to be missing.

File: target/linux/ramips/image/mt7621.mk [hunk 21]
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
 endef
 TARGET_DEVICES += zio_freezio
 
+define Device/zyxel_keenetic-giga-iii
+  $(Device/dsa-migration)
+  KERNEL_SIZE := 4096k  # kernel partition size
+  IMAGE_SIZE := 50m
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1691]: Please specify in "k" instead, thats common practice.

File: target/linux/ramips/image/mt7621.mk [hunk 22]
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
...
+  BLOCKSIZE := 128k     # NAND erase size (block size)
+  PAGESIZE := 2048      # NAND page size
+  DEVICE_VENDOR := ZyXEL
+  DEVICE_MODEL := Keenetic Giga III
+  SUPPORTED_DEVICES += kng_re
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1696]: This line should be dropped, unless it's needed for factory flashing.

File: target/linux/ramips/image/mt7621.mk [hunk 23]
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
...
+  DEVICE_PACKAGES := kmod-mt76x2 kmod-usb2 kmod-usb3
+  DEVICE_PACKAGES += kmod-usb-ledtrig-usbport kmod-usb-storage
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1698]: Please use a multiline assignment for one variable like all the others.
I don't think usb-storage should be included by default.

File: target/linux/ramips/image/mt7621.mk [hunk 24]
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
...
+  VER_NUM := $$(shell cut -d '-' -f 1 <<< "$$(VERSION_NUMBER)" )
+  VER_DATE := $$(shell date --utc -d @$$(SOURCE_DATE_EPOCH) +%y%m%d)
+  VER_FULL := $$(VER_NUM) $$(VER_DATE)
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | uImage none -n 'KN-GIGA3 $$(VERSION_DIST) $$(VER_FULL)'
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1702]: please create a Build recipe for that instead of introducing local variables here.

File: target/linux/ramips/image/mt7621.mk [hunk 25]
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
...
+  UBINIZE_OPTS := -E 5   # EOD marks to "hide" factory sig at EOF
+  IMAGES := sysupgrade.tar
+  IMAGE/sysupgrade.tar := sysupgrade-tar | append-metadata
+  IMAGES += factory.bin
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1706]: IMAGES in one line.

File: target/linux/ramips/image/mt7621.mk [hunk 26]
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
...
+  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi
+  IMAGE/factory.bin += | check-size | pad-to $$(BLOCKSIZE)
+  IMAGE/factory.bin += | zyimage -d 10368 -v "$$(DEVICE_VENDOR) $$(DEVICE_MODEL) $$(VERSION_DIST) $$(VER_FULL)"
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1709]: One multiline command. As above, please create a Build recipe for the last step.

File: target/linux/ramips/mt7621/base-files/etc/board.d/01_leds [hunk 27]
```diff
@@ -122,6 +122,11 @@ xiaomi,redmi-router-ac2100)
 youhua,wr1200js)
 	ucidef_set_led_netdev "internet" "INTERNET" "green:wan" "wan"
 	;;
+zyxel,keenetic-giga-iii)
+	ucidef_set_led_netdev  "internet" "internet"  "green:internet" "wan"
+	ucidef_set_led_netdev  "wifi5g"   "Wifi 5G"   "green:wifi5G"   "wlan0"
+	ucidef_set_led_netdev  "wifi2g"   "Wifi 2.4G" "green:wifi2G"   "wlan1"
+	;;
```
Comment (adschm) [inline target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:129]: Please use the same formatting as anybody else.

Wifi setup can be dropped here since you have DT triggers.

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 28]
```diff
@@ -0,0 +1,301 @@
...
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "zyxel,keenetic-giga-iii", "mediatek,mt7621-soc";
+	model = "ZyXEL Keenetic Giga III";
+
+	aliases {
+		label-mac-device = &gmac0;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:11]: This should be &wan (you should add the DT label as described below).

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts [hunk 29]
```diff
@@ -0,0 +1,301 @@
...
+			status = "okay";
+			label = "lan2";
+		};
+
+		port@3 {
+			status = "okay";
+			label = "lan1";
+		};
+
+		port@4 {
```
Comment (adschm) [inline target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:269]: Use "wan: port@4 {" here, so you can use the mac address for label-mac-device.

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network [hunk 30]
```diff
@@ -157,6 +157,11 @@ ramips_setup_macs()
 		wan_mac=$label_mac
 		lan_mac=$(macaddr_add $label_mac 1)
 		;;
+	zyxel,keenetic-giga-iii)
+		lan_mac=$(mtd_get_mac_binary factory 0x4)
+		wan_mac=$(mtd_get_mac_binary factory 0x28)
+		label_mac=$wan_mac
+		;;
```
Comment (adschm) [inline target/linux/ramips/mt7621/base-files/etc/board.d/02_network:164]: This can be dropped entirely when you've fixed your DTS.

**Review by adschm — CHANGES_REQUESTED** [review]

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts (hunk 1 above)
Comment (remittor) [reply target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:3]: fixed

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts (hunk 2 above)
Comment (remittor) [reply target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts]: fixed

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts (hunk 3 above)
Comment (remittor) [reply target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts]: fixed

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts (hunk 4 above)
Comment (remittor) [reply target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts]: fixed

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts (hunk 5 above)
Comment (remittor) [reply target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:39]: fixed

File: target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts (hunk 19 above)
Comment (remittor) [reply target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:288]: Although not currently used, it should be kept for information.

File: target/linux/ramips/image/mt7621.mk (hunk 21 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1691]: Megabytes are also used:
```
$ grep -r --include "*.mk" -E "IMAGE_SIZE := [0-9]{1,3}m"
layerscape/image/armv8_64b.mk:  IMAGE_SIZE := 64m
layerscape/image/armv7.mk:  IMAGE_SIZE := 64m
ath79/image/nand.mk:  IMAGE_SIZE := 116m
```
And this suffix is handled correctly: https://github.com/openwrt/openwrt/blob/ac03e246351fbfa98e98f0319cce3a8663f836d3/include/image-commands.mk#L171

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1697]: This variable `PAGESIZE` must be specified as well as the variable `UBINIZE_OPTS := -E 5`.
See: https://github.com/openwrt/openwrt/blob/ac03e246351fbfa98e98f0319cce3a8663f836d3/include/image-commands.mk#L101-L114
See ubinize: https://github.com/sigma-star/mtd-utils/blob/f0b32c7ba5136151a961c3a43c8ff430b33d7c33/ubi-utils/ubinize.c#L47-L48
If we use a method `append-ubi`, then we must specify the variable `PAGESIZE`.

File: target/linux/ramips/image/mt7621.mk (hunk 24 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk]: I think this is completely superfluous.
Did you talk about something similar?
```
define Device/repo-version
  REPO_VER := $$(shell cut -d '-' -f 1 <<< "$$(VERSION_NUMBER)" )
  REPO_DATE := $$(shell date --utc -d @$$(SOURCE_DATE_EPOCH) +%y%m%d)
  REPO_VER_DATE := $$(REPO_VER) $$(REPO_DATE)
  REPO_FULL_VER := $$(VERSION_DIST) $$(REPO_VER) $$(REPO_DATE)
endif
```

File: target/linux/ramips/image/mt7621.mk (hunk 21 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:1691]: Out of multiple hundred uses of this variable, there is indeed a one-digit number of cases, on other targets, that does not use "k". That's essentially what common practice means.
So, please change it after you have proven my point now.

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:1697]: Funny, how you discuss all of the variables here except the line I was referring to.
I was referring to the line directly before my comment, which is SUPPORTED_DEVICES ...

File: target/linux/ramips/image/mt7621.mk (hunk 24 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk]: No, this is a common device definition. I'm talking about a build recipe, these are the blocks starting with "Build/", i.e. those you are using when you put all these nice terms like "kernel-bin" or "loader-kernel" ...

Comment (adschm) [reply target/linux/ramips/image/mt7621.mk]: Some thing like this (just typed it without too much care):

```
define Build/my_special_uimage_command
  $(eval repo_ver=$$(shell cut -d '-' -f 1 <<< "$$(VERSION_NUMBER)"))
  $(eval repo_date=$$(shell date --utc -d @$$(SOURCE_DATE_EPOCH) +%y%m%d))
  uImage none -n '$(word 1, $(1)) $$(VERSION_DIST) $(repo_ver) $(repo_date)'
endef

[...]
KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | my_special_uimage_command KN-GIGA3
[...]
```

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1697]: You will be surprised when you enter the command:
`grep -r --include "*.mk" -n -E "SUPPORTED_DEVICES +" | wc -l`
I doubt very much that all found devices need to specify this value for factory flashing.

File: target/linux/ramips/image/mt7621.mk (hunk 24 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk]: Maybe you can then do what you like with a separate commit?
I really think that making a specially separate recipe for one device is a complete overkill.

File: target/linux/ramips/image/mt7621.mk (hunk 21 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1691]: OK. So I'll write the value 51000k

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1697]: grep finds several Keenetic:
```
ramips/image/mt7620.mk:1329:  SUPPORTED_DEVICES += kn_rc
ramips/image/mt7620.mk:1342:  SUPPORTED_DEVICES += kn_rf
ramips/image/mt7620.mk:1356:  SUPPORTED_DEVICES += kng_rc
```

File: target/linux/ramips/image/mt7621.mk (hunk 24 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk]: And in which file should this recipe be placed?
Is it possible in `target/linux/ramips/image/Makefile` ?

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:1697]: Have you ever considered the possibility that I know what I'm speaking of and instead of you might actually know what SUPPORTED_DEVICES _is_ good for?

File: target/linux/ramips/image/mt7621.mk (hunk 21 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:1691]: You have no clue what this value does, do you?

File: target/linux/ramips/image/mt7621.mk (hunk 24 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk]: This depends on the devices it will be used for. If it's just for one device, you can put in the same file as the device definition (of course, at the top and keeping proper sorting of recipes).

**Comment by adschm** [comment]
I'm sorry for being a bit rude here. But you obviously try to teach me on things that you don't really understand yourself, and that's not helping. Note that I have commit access to this repository, so you can except that when I give you advice on how to fix things, there is a high chance that my pointer is correct.

There is a lot of submissions to review and I don't have time for arguments like that, so please, again, excuse my direct address here.

**Comment by remittor** [comment]
> I'm sorry for being a bit rude here.

It's OK. You have found many bugs.

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1697]: Я прежде всего доверяю написанному коду и скриптам.
The «kng_re» value is stored in the bootloader configs ("U-Config" partition).
```
servicehost=ndss.11.zyxel.ndmsystems.com
servicepass=6Wg_____480
sernumb=S150_____22
ndmhwid=kng_re
servicetag=500______776
wlanssid=Keenetic-8204
wlankey=S3i___HZ
wlanwps=658___88
country=RU
```

File: target/linux/ramips/image/mt7621.mk (hunk 21 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1691]: Do you want me to explain to you what this code does?
https://github.com/openwrt/openwrt/blob/ac03e246351fbfa98e98f0319cce3a8663f836d3/include/image-commands.mk#L169-L176
I have not found the use of this variable anywhere else.

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:1697]: Interesting. Now, would you please remove the SUPPORTED_DEVICES line?

SUPPORTED_DEVICES is for upgrades between different OpenWrt firmwares only and since this is the first OpenWrt firmware (unless OEM firmware is, which is why I asked the initial question), we don't need a custom second value here.

File: target/linux/ramips/image/mt7621.mk (hunk 21 above)
Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:1691]: "check-size" checks the size of the final image. For devices with small flash, typically 4M/8M, you will quickly manage to create images that are bigger than the available amount of firmware partition space to write them to. Since that might result in a broken or even unbootable device, the build routine checks the image size and will abort the build process if the "image" is too large when reaching the check-size check.

File: target/linux/ramips/image/mt7621.mk (hunk 22 above)
Comment (remittor) [reply target/linux/ramips/image/mt7621.mk:1697]: OK. I will delete it. Looking at the code, I realized that it has a double meaning (not only about what you said).

**Comment by remittor** [comment]
ping @adschm

**Comment by nm17** [comment]
@remittor Any plans on continuing the development?

**Comment by remittor** [comment]
@nm17, So it's all done. You can merge and build the firmware.

**Comment by jospezial** [comment]
You did not exchange the telnet advise with ssh.
https://openwrt.org/toh/zyxel/zyxel_keenetic_omni_rev._a#oem_easy_installation should get an update to ssh.

**Comment by nm17** [comment]
@remittor why is the pr not passing the review?

**Comment by neheb** [comment]
Must be rebased against current master.

**Comment by remittor** [comment]
> Must be rebased against current master.

Done

**Comment by remittor** [comment]
> Two changes

Why two changes?

> GPIO switch LED support. The Netgear wax202 PR has an example.

Should I add to the `01_leds` file using the `ucidef_set_led_usbport`?
No device based on mt7621 uses this specifier.

**Comment by neheb** [comment]
Then no.

I'm talking about switch LEDs, not normal GPIO ones.

**Comment by arinc9** [comment]
> > Implement the changes in #10238
> 
> Added this as a separate commit. But I haven't tested these changes.

The commit log is too vague to be accepted. The main commit isn't merged yet anyway, so just squash em.

**Comment by remittor** [comment]
> Implement the changes in #10238

Added this patch. I also tested the work with these changes (without testing the maximum throughput).

**Comment by KashBrus** [comment]
Hi @remittor any chance or time gap when it would be merged?
Will it be available from official builds?

File: target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh [hunk 31]
```diff
@@ -120,6 +120,21 @@ platform_do_upgrade() {
...
+		if [ "$magic" == "27051956" ]; then
+			echo "Forced upgrade firmware partition..." 
+			mtd -f write $1 firmware
+			echo "uImage-firmware write successful! Reboot..."
+			sync
+			umount -a
+			reboot -f
+			exit 1
+		fi
+		CI_KERNPART="kernel"
```
Comment (hauke) [inline target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:135]: `CI_KERNPART="kernel"` is already the default. No need to set it here.

File: target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh [hunk 32]
```diff
@@ -120,6 +120,21 @@ platform_do_upgrade() {
 	ubnt,edgerouter-x-sfp)
 		platform_upgrade_ubnt_erx "$1"
 		;;
+	zyxel,keenetic-giga-iii)
+		local magic=$( nand_get_magic_long "$1" 0 )
+		echo "Firmware image $1 contain magic $magic"
+		if [ "$magic" == "27051956" ]; then
```
Comment (hauke) [inline target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:126]: Why is this special handling needed here?
Please add a comment.

File: target/linux/ramips/image/mt7621.mk [hunk 33]
```diff
@@ -2274,6 +2274,24 @@ define Device/zio_freezio
 endef
 TARGET_DEVICES += zio_freezio
 
+define Device/zyxel_keenetic-giga-iii
+  $(Device/dsa-migration)
+  KERNEL_SIZE := 4096k  # kernel partition size
+  IMAGE_SIZE := 51000k
+  BLOCKSIZE := 128k     # NAND erase size
+  PAGESIZE := 2048      # NAND page size
+  UBINIZE_OPTS := -E 5  # EOD marks to "hide" factory sig at EOF
```
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:2283]: Please replace these nand common definitions with `$(Device/nand)`.
ref: https://github.com/openwrt/openwrt/commit/b09a83808a3a4f6ff2ff0cf1f12fdc4193357896

**Comment by Nekotekina** [comment]
Why did I have such a bad luck choosing a router that isn't officially supported...
Nevertheless I installed openwrt with this PR patch and it seems working, more or less. Thank you.

**Comment by GoetzGoerisch** [comment]
Thank you for your PR.

Just a heads up, we had a recent [treewide change from ZyXEL to Zyxel](https://github.com/openwrt/openwrt/commit/3774f3272ee976b922306a881b6eb603e79f81ed).
Please adjust your commit message and code accordingly. Thank you!
